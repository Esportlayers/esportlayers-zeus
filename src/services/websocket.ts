import { WebsocketInstance } from "..";

export function sendMessage(userId: number, type: string, value: any): void {
  const connections = [...WebsocketInstance.getWss().clients.values()].filter(
    (c) =>
      //@ts-ignore
      c.user &&
      //@ts-ignore
      c.user.id === userId &&
      //@ts-ignore
      (!type.includes("gsi") || !c.noGsiEvents)
  );
  connections.forEach((client) => client.send(JSON.stringify({ type, value })));
}
