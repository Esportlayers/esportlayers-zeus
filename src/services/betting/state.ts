import { User } from "@streamdota/shared-types";
import dayjs from "dayjs";
import { getObj, setObj } from "../../loader/redis";
import { requireBetOverlay } from "../entity/BetOverlay";
import { createBet, createBetRound, getRoundId, patchBetRound } from "../entity/BetRound";
import { getUserByTrustedChannel, loadUserById } from "../entity/User";
import { fetchChatterCount } from "../twitchApi";
import { publish } from "../twitchChat";
import { sendMessage } from "../websocket";
import { getBettingCommands } from "./chatCommands";

interface BetRoundData {
    status: 'stream_delay' | 'betting' | 'game_running' | 'finished';
    overlayVisibleUntil: number;
    overlayVisible: boolean;
    streamDelay: number;
    votingStartingAt: number;
    votingTimeRemaining: number;
    votingPossibleUntil: number;
    voteCreated: number;
    totalVotesCount: number;
    chatterCounts: number;
    teamACount: number;
    teamAVoters: string[];
    teamBCount: number;
    teamBVoters: string[];
    allVoters: string[];
    winner: null | string;
    winnerAnnouncement: null | number;
    announcedStart: boolean;
    announcedVoteEnd: boolean;
    announcedWinner: boolean;
}

const channeUserCache = new Map<string, User>();

export async function requireUser(channel: string): Promise<User> {
    const lowerChannel = channel.toLowerCase();
    if(!channeUserCache.has(lowerChannel)) {
        const {id} = await getUserByTrustedChannel(channel);
        const user = (await loadUserById(id))!;
        channeUserCache.set(lowerChannel, user);
    }
    return channeUserCache.get(lowerChannel)!;
}


function activeVoteKeys(): string {
    return `betting_active_channels`;
}

async function addChannel(channel: string): Promise<void> {
    const channels = (await getObj<string[]>(activeVoteKeys()) || []);
    await setObj(activeVoteKeys(), channels.concat(channel));
}

async function removeChannel(channel: string): Promise<void> {
    const channels = (await getObj<string[]>(activeVoteKeys()) || []);
    await setObj(activeVoteKeys(), channels.filter((c) => c !== channel));
}

function roundKey(channel: string): string {
    return `betting_round_state_${channel}`;
}

async function updateListener(channel: string, userId: number): Promise<void> {
    const currentRound = await getObj<BetRoundData>(roundKey(channel));
    sendMessage(userId, 'betting_v2', currentRound);
}

async function startVote(channel: string, user: User): Promise<void> {
    const currentRound = await getObj<BetRoundData>(roundKey(channel));
    const {startBet: startBetCommand, bet: betCommand } = await getBettingCommands(channel);
    if(startBetCommand) {
        let message = startBetCommand.message.replace(/\{BET_COMMAND\}/g, betCommand?.command || '');
        message = message.replace(/\{TEAM_A\}/g, user?.teamAName || 'a');
        message = message.replace(/\{TEAM_B\}/g, user?.teamBName || 'b');
        publish(channel, message);
        await createBetRound(user.id, user.betSeasonId);
        await setObj(roundKey(channel), {
            ...currentRound,
            status: 'betting',
            announcedStart: true,
        })
        await updateListener(channel, user.id);
    }
}

async function finishVote(channel: string, user: User): Promise<void> {
    const roundId = await getRoundId(user.id);
    await patchBetRound(roundId, {status: 'running'});
    publish(channel, 'Die Votes sind geschlossen.');
    const currentRound = await getObj<BetRoundData>(roundKey(channel));
    await setObj(roundKey(channel), {
        ...currentRound,
        status: 'game_running',
        announcedVoteEnd: true,
    })
    await updateListener(channel, user.id);
}

async function closeVote(channel: string, user: User): Promise<void> {
    const roundId = await getRoundId(user.id);
    const currentRound = await getObj<BetRoundData>(roundKey(channel));
    await patchBetRound(roundId, {result: currentRound!.winner!, status: 'finished'}, true, user.id);
    const {winner: winnerCommand} = await getBettingCommands(channel);
    publish(channel, winnerCommand.message.replace(/\{WINNER\}/g, currentRound!.winner!));
    await setObj(roundKey(channel), null);
    await removeChannel(channel);
    await updateListener(channel, user.id);
}

export async function initializeBet(channel: string, userId: number): Promise<void> {
    const user = await loadUserById(userId);
    const currentRound = await getObj<BetRoundData>(roundKey(channel));
    
    if(user && (! currentRound || currentRound.status === 'finished')) {
        const overlay = await requireBetOverlay(userId);
        const chatters = await fetchChatterCount(channel.substring(1));
        const ts = dayjs().unix();
        await setObj(roundKey(channel), {
            status: 'stream_delay',
            overlayVisibleUntil: ts + overlay.timerDuration,
            overlayVisible: true,
            streamDelay: user.streamDelay,
            votingStartingAt: ts + user.streamDelay,
            votingTimeRemaining: ts + user.streamDelay,
            votingPossibleUntil: ts + user.streamDelay + overlay.timerDuration,
            voteCreated: ts,
            totalVotesCount: 0,
            chatterCounts: chatters,
            teamACount: 0,
            teamBCount: 0,
            teamAVoters: [],
            teamBVoters: [],
            allVoters: [],
            winner: null,
            winnerAnnouncement: null,
            announcedStart: false,
            announcedVoteEnd: false,
            announcedWinner: false,
        });
        await addChannel(channel);
        await updateListener(channel, userId);
    }
}

export async function registerBet(channel: string, userId: number, votingUserId: number, displayName: string, userName: string, message: string): Promise<void> {
    const currentRound = await getObj<BetRoundData>(roundKey(channel));
    if(currentRound && currentRound.status === 'betting' && !currentRound?.allVoters.includes(userName)) {
        const {bet} = await getBettingCommands(channel);
        const payload = message.substr(bet.command.length + 1).toLowerCase();
        const user = await loadUserById(userId);
        const betOnTeamA = payload.toLowerCase() === user?.teamAName.toLowerCase();
        const betOnTeamB = payload.toLowerCase() === user?.teamBName.toLowerCase();

        if(user && (betOnTeamA || betOnTeamB)) {
            await createBet(userId, votingUserId, displayName, userName, betOnTeamA ? user.teamAName : user.teamBName);
            await setObj(roundKey(channel), {
                ...currentRound,
                allVoters: currentRound.allVoters.concat(userName),
                ...(betOnTeamA ? {teamAVoters: currentRound.teamAVoters.concat(userName), teamACount: currentRound.teamACount + 1, totalVotesCount: currentRound.totalVotesCount + 1} : {}),
                ...(betOnTeamB ? {teamBVoters: currentRound.teamBVoters.concat(userName), teamBCount: currentRound.teamBCount + 1, totalVotesCount: currentRound.totalVotesCount + 1} : {}),
            });

            await updateListener(channel, userId);
        }
    }
}

export async function resolveBet(channel: string, userId: number, result: string): Promise<void> {
    const currentRound = await getObj<BetRoundData>(roundKey(channel));
    if(currentRound && currentRound.status === 'game_running') {
        const user = await loadUserById(userId);
        if(user?.teamAName.toLowerCase() === result.toLowerCase() || user?.teamBName.toLowerCase() === result.toLowerCase()) {
            const ts = dayjs().unix();
            if(user) {
                await setObj(roundKey(channel), {
                    ...currentRound,
                    winner: user?.teamAName.toLowerCase() === result.toLowerCase() ? user?.teamAName : user?.teamBName,
                    winnerAnnouncement: ts + user?.streamDelay,
                });

                await updateListener(channel, userId);
            }
        }
    }
}

async function streamDelayChecker(data: BetRoundData, ts: number, channel: string, user: User): Promise<void> {
    if(data.overlayVisible && data.overlayVisibleUntil <= ts) {
        await setObj(roundKey(channel), {...data, overlayVisible: false});
        await updateListener(channel, user.id);
    }

    if(data.votingStartingAt <= ts && !data.announcedStart) {
        await startVote(channel, user);
    }
}

async function bettingChecker(data: BetRoundData, ts: number, channel: string, user: User): Promise<void> {
    if(data.votingPossibleUntil <= ts && !data.announcedVoteEnd) {
        await finishVote(channel, user);
    } else {
        await setObj(roundKey(channel), {
            ...data,
            votingTimeRemaining: data.votingPossibleUntil - ts,
        });
        await updateListener(channel, user.id);
    }
}

async function gameRunningChecker(data: BetRoundData, ts: number, channel: string, user: User): Promise<void> {
    if(data.winner && data.winnerAnnouncement! <= ts && !data.announcedWinner) {
        await closeVote(channel, user);
    }
}

const statusChecker = {
    stream_delay: streamDelayChecker,
    betting: bettingChecker,
    game_running: gameRunningChecker,
}

export async function updateBetRounds(): Promise<void> {
    const activeBettingChannel = await getObj<string[]>(activeVoteKeys()) || [];
    const ts = dayjs().unix();

    for(const channel of activeBettingChannel) {
        const round = await getObj<BetRoundData>(roundKey(channel));
        const user = await requireUser(channel);

        if(round && round.status !== 'finished') {
            await statusChecker[round.status](round, ts, channel, user);
        }
    }
}

export async function newBettingListener(userId: number, displayName: string): Promise<void> {
    await updateListener('#' + displayName.toLowerCase(), userId);

}