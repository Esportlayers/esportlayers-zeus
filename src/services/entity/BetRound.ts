import { loadUserById } from "./User";
import { getConn } from "../../loader/db";
import { RowDataPacket } from "mysql2";
import { requireWatcher } from "./Watcher";
import {BetRound, BetRoundStats} from '@streamdota/shared-types';
import { fetchChatterCount } from "../twitchApi";
import { updateBetState } from "../betting/state";

async function getRound(userId: number): Promise<number> {
    const user = await loadUserById(userId);
    if(user && user.betSeasonId) {
        const conn = await getConn();
        const [roundRows] = await conn.execute<Array<{round: number} & RowDataPacket>>('SELECT round FROM bet_rounds WHERE bet_season_id = ? ORDER BY round DESC LIMIT 1', [user.betSeasonId]);
        await conn.end();

        return roundRows.length > 0 ? roundRows[0].round : 0;
    }

    return 0;
}

export async function getRoundId(userId: number): Promise<number> {
    const user = await loadUserById(userId);
    if(user && user.betSeasonId) {
        const conn = await getConn();
        const [roundRows] = await conn.execute<Array<{id: number} & RowDataPacket>>('SELECT id FROM bet_rounds WHERE bet_season_id = ? AND user_id = ? ORDER BY round DESC LIMIT 1', [user.betSeasonId, userId]);
        await conn.end();

        return roundRows.length > 0 ? roundRows[0].id : 0;
    }

    return 0;
}

interface DecoratedBetRound extends BetRound {
    created: number;
    total: number;
    aBets: string;
    bBets: string;
}

export async function getRoundById(roundId: number): Promise<DecoratedBetRound | null> {
    const conn = await getConn();
    const [rows] = await conn.execute<Array<DecoratedBetRound & RowDataPacket>>(`
        SELECT br.id, 
               br.bet_season_id as betSeason, 
               br.round, 
               br.status, 
               br.result,
               br.chatters as chatters,
               UNIX_TIMESTAMP(br.created) as created,
               COUNT(b.id) as total,
               SUM(IF(b.bet = 'a', 1, 0)) as aBets,
               SUM(IF(b.bet = 'b', 1, 0)) as bBets
        FROM bet_rounds br
   LEFT JOIN bets b ON b.bet_round_id = br.id
       WHERE br.id = ? 
    GROUP BY br.id`, [roundId]);
    await conn.end();

    return rows.length > 0 ? rows[0] : null;
}


export async function getBetSeasonRounds(seasonId: number): Promise<BetRoundStats[]> {
    const conn = await getConn();
    const [rows] = await conn.execute<Array<BetRoundStats & RowDataPacket>>(`
        SELECT 
            br.id, 
            br.round, 
            br.status, 
            br.result, 
            br.user_id as userId,
            br.chatters as chatters,
            u.display_name as displayName,
            UNIX_TIMESTAMP(br.created) as created,
            COUNT(b.id) as total,
            SUM(IF(b.bet = 'a', 1, 0)) as aBets,
            SUM(IF(b.bet = 'b', 1, 0)) as bBets,
            br.bet_season_id as betSeason
       FROM bet_rounds br
  LEFT JOIN bets b ON b.bet_round_id = br.id
 INNER JOIN user u ON u.id = br.user_id
      WHERE br.bet_season_id = ?
   GROUP BY br.id
   ORDER BY br.created DESC
`, 
        [seasonId]
    );
    await conn.end();

    return rows;
}

export async function createBetRound(userId: number, seasonId: number | null, notify: boolean = false): Promise<void> {
    if(seasonId) {
        const conn = await getConn();
        const round = (await getRound(userId)) + 1;
        const {displayName} = (await loadUserById(userId))!;
        const chatters = await fetchChatterCount(displayName);

        await conn.execute(
            'INSERT INTO bet_rounds (id, bet_season_id, user_id, round, created, status, result, chatters) VALUES (NULL, ?, ?, ?, NOW(), ?, "", ?)', 
            [seasonId, userId, round, 'betting', chatters]
        );
        await conn.end();    
    }
    
    if(notify) {
        await updateBetState(userId, true);
    }
}

interface PatchableData {
    status: 'betting' | 'running' | 'finished';
    result: string;
}

export async function patchBetRound(roundId: number, data: Partial<PatchableData>, notify: boolean = false, userId: number | null = null): Promise<void> {
const conn = await getConn();

    if(data.result) {
        await conn.execute('UPDATE bet_rounds SET result=? WHERE id=?', [data.result, roundId]);
    }

    if(data.status) {
        await conn.execute('UPDATE bet_rounds SET status=? WHERE id=?', [data.status, roundId]);
    }

    await conn.end();

    if(notify) {
        await updateBetState(userId!, false, data?.status === 'finished');
    }
}

export async function deleteBetRound(roundId: number): Promise<void> {
    const conn = await getConn();
    await conn.execute('DELETE FROM bets WHERE bet_round_id=?', [roundId]);
    await conn.execute('DELETE FROM bet_rounds WHERE id=?', [roundId]);
    await conn.end();
}

export async function createBet(userId: number, twitchId: number, displayName: string, username: string, bet: string): Promise<void> {
    const conn = await getConn();
    const watcher = await requireWatcher(twitchId, displayName, username, userId);
    const betRound = await getRoundId(userId);

    await conn.execute(
        'INSERT INTO bets (id, watcher_id, bet_round_id, created, bet) VALUES (NULL, ?, ?, NOW(), ?)',
        [watcher.id, betRound, bet]
    );
    await conn.end();
}