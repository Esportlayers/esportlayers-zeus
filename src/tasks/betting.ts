import { green } from "chalk";
import interval from "interval-promise";
import { updateBetRounds } from "../services/betting/state";

console.log(green('ðŸ’° Registered betting update task'));
interval(async () => updateBetRounds(), 1000);