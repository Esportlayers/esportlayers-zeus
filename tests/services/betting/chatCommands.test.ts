const fakeCommands = [
    {active: true, type: 'betting_user', command: '!a', message: 'a response'},
    {active: false, type: 'betting_user', command: '!b', message: 'b response'},
    {active: true, type: 'betting_streamer', command: '!c', message: 'c response'},
    {active: true, type: 'betting_streamer', command: '!e', message: 'e response', identifier: 'startbet'},
    {active: true, type: 'betting_user', command: '!f', message: 'f response', identifier: 'bet'},
    {active: true, type: 'betting_streamer', command: '!g', message: 'g response', identifier: 'betwinner'},
    {active: true, type: 'default', command: '!d', message: 'd response'},
];

const fakeToplist = [
    {name: 'a', won: 4, total: 4},
    {name: 'b', won: 3, total: 4},
    {name: 'c', won: 3, total: 3},
]

jest.mock('../../../src/services/entity/Command', () => ({
    getUserCommands: jest.fn(async () => fakeCommands)
}));

jest.mock('../../../src/services/entity/BetSeasons', () => ({
    getUserSeasonStats: jest.fn(async () => ({won: 4, total: 6})),
    seasonTopList: jest.fn(async () => fakeToplist),
}));

jest.mock('../../../src/services/entity/BetRound', () => ({
    createBetRound: jest.fn(),
    createBet: jest.fn(),
    getRoundId: jest.fn(async () => 1),
    patchBetRound: jest.fn(),
}));

jest.mock('../../../src/services/websocket', () => ({
    sendMessage: jest.fn(),
}));

jest.mock('../../../src/services/betting/state', () => ({
    requireBettingRound: jest.fn(),
    requireUser: jest.fn(),
    startBet: jest.fn(),
}));

process.env.SENTRY_DSN = '';

import {
    requireUserCommands,
    clearBettingCommandsCache, 
    replaceBetPlaceholder, 
    handleStaticCommands, 
    getBettingCommands,
    handleUserBet,
    processCommands
} from '../../../src/services/betting/chatCommands';
import {getUserCommands} from '../../../src/services/entity/Command';
import { User } from '@streamdota/shared-types';
import { ChatUserstate } from 'tmi.js';
import { CurrentBetRound, startBet } from '../../../src/services/betting/state';
import { requireBettingRound, requireUser } from '../../../src/services/betting/state';
import { sendMessage } from '../../../src/services/websocket';
import { patchBetRound, createBet } from '../../../src/services/entity/BetRound';

const fakeChannel = '#test';

describe('requireUserCommands', () => {
    test('test first call without store', async () => {
        const commands = await requireUserCommands(fakeChannel , 1);
        expect(getUserCommands).toHaveBeenCalledTimes(1);
        expect(getUserCommands).toHaveBeenCalledWith(1);
        expect(commands).toMatchSnapshot();
    });

    test('test second call with store', async () => {
        (getUserCommands as jest.Mock).mockClear();
        const commands = await requireUserCommands(fakeChannel , 1);
        expect(getUserCommands).toHaveBeenCalledTimes(0);
        expect(commands).toMatchSnapshot();
    });

    test('clear of store', async () => {
        (getUserCommands as jest.Mock).mockClear();
        clearBettingCommandsCache(fakeChannel + '1');
        expect(getUserCommands).toHaveBeenCalledTimes(0);
        clearBettingCommandsCache(fakeChannel);
        expect(getUserCommands).toHaveBeenCalledTimes(0);

        const commands = await requireUserCommands(fakeChannel , 1);
        expect(commands).toMatchSnapshot();
    })
})

describe('replaceBetPlaceholder', () => {
    test('{USER}', async () => {
        const replaced = await replaceBetPlaceholder('Hallo {USER}', 'GriefCode', 1);
        expect(replaced).toEqual('Hallo GriefCode');
    })
    test('{TOPLIST_STATS}', async () => {
        const replaced = await replaceBetPlaceholder('Toplist {TOPLIST_STATS}', 'GriefCode', 1);
        expect(replaced).toEqual('Toplist 1. a (4/4)  2. b (3/4)  3. c (3/3)');
    })
    test('{USER_BETS_CORRECT}', async () => {
        const replaced = await replaceBetPlaceholder('Correct {USER_BETS_CORRECT}', 'GriefCode', 1);
        expect(replaced).toEqual('Correct 4');
    })
    test('{USER_BETS_WRONG}', async () => {
        const replaced = await replaceBetPlaceholder('Wrong {USER_BETS_WRONG}', 'GriefCode', 1);
        expect(replaced).toEqual('Wrong 2');
    })
    test('{USER_BETS_TOTAL}', async () => {
        const replaced = await replaceBetPlaceholder('Total {USER_BETS_TOTAL}', 'GriefCode', 1);
        expect(replaced).toEqual('Total 6');
    })
    test('{USER_BETS_ACCURACY}', async () => {
        const replaced = await replaceBetPlaceholder('Accuracy {USER_BETS_ACCURACY}', 'GriefCode', 1);
        expect(replaced).toEqual('Accuracy 66%');
    })
});

describe('handleStaticCommands', () => {
    const fakeUser =  {id: 1, betSeasonId: 1} as unknown as User;
    test('static non betting command', async () => {
        const res = await handleStaticCommands(fakeChannel, '!d', fakeUser, 'GriefCode');
        expect(res).toBeFalsy();
    });
    test('static betting command', async () => {
        const res = await handleStaticCommands(fakeChannel, '!a', fakeUser, 'GriefCode');
        expect(res).toBeTruthy();
    });
    test('static betting non active command', async () => {
        const res = await handleStaticCommands(fakeChannel, '!b', fakeUser, 'GriefCode');
        expect(res).toBeFalsy();
    });
})

describe('getBettingCommands', () => {
    test('recv', async () => {
        const commands = await getBettingCommands(1, fakeChannel);
        expect(commands).toMatchSnapshot();
    })
})

const fakeTags = {
    'user-id': 123,
    'display-name': 'GriefCode',
    'username': 'griefcode'
} as unknown as ChatUserstate;

describe('handleUserBet', () => {
    const getFakeRound = () => ({
        bets: 4,
        aBets: 2,
        bBets: 2,
        betters: ['a', 'b', 'c', 'd']
    } as unknown as CurrentBetRound);
    
    test('bet a', async () => {
        const fakeRound = getFakeRound();
        const {bet} = await getBettingCommands(1, fakeChannel);
        await handleUserBet('!f a', bet, fakeTags, fakeRound, 1)
        expect(fakeRound.aBets).toEqual(3);
        expect(fakeRound.bets).toEqual(5);
        expect(fakeRound.betters.includes('griefcode')).toBeTruthy();
    })
    test('bet b', async () => {
        const fakeRound = getFakeRound();
        const {bet} = await getBettingCommands(1, fakeChannel);
        await handleUserBet('!f b', bet, fakeTags, fakeRound, 1)
        expect(fakeRound.bBets).toEqual(3);
        expect(fakeRound.bets).toEqual(5);
        expect(fakeRound.betters.includes('griefcode')).toBeTruthy();
    })
    test('invlaid bet', async () => {
        const fakeRound = getFakeRound();
        const {bet} = await getBettingCommands(1, fakeChannel);
        await handleUserBet('!f c', bet, fakeTags, fakeRound, 1)
        expect(fakeRound.aBets).toEqual(2);
        expect(fakeRound.bBets).toEqual(2);
        expect(fakeRound.bets).toEqual(4);
        expect(fakeRound.betters.includes('griefcode')).toBeFalsy();
    })
})

describe('processCommands', () => {
    beforeEach(() => {
        (requireUser as jest.Mock).mockClear();
        (requireBettingRound as jest.Mock).mockClear();
        (sendMessage as jest.Mock).mockClear();
        (startBet as jest.Mock).mockClear();
        (patchBetRound as jest.Mock).mockClear();
        (createBet as jest.Mock).mockClear();
    })

    test('unknown user', async () => {
        (requireUser as jest.Mock).mockReturnValueOnce(null);
        await processCommands(fakeChannel, fakeTags, '!e');
        expect(requireBettingRound).not.toHaveBeenCalled();
    })
    test('no bets used', async () => {
        (requireUser as jest.Mock).mockReturnValueOnce({useBets: false});
        await processCommands(fakeChannel, fakeTags, '!e');
        expect(requireBettingRound).not.toHaveBeenCalled();
    })
    test('static command', async () => {
        (requireUser as jest.Mock).mockReturnValueOnce({useBets: true});
        await processCommands(fakeChannel, fakeTags, '!a');
        expect(requireBettingRound).not.toHaveBeenCalled();
    })
    test('unknown command', async () => {
        (requireUser as jest.Mock).mockReturnValueOnce({useBets: true});
        await processCommands(fakeChannel, fakeTags, '!X');
        expect(requireBettingRound).not.toHaveBeenCalled();
    })
    test('startbet from non broadcaster', async () => {
        (requireUser as jest.Mock).mockReturnValueOnce({useBets: true});
        await processCommands(fakeChannel, fakeTags, '!e');
        expect(requireBettingRound).toHaveBeenCalledTimes(1);
        expect(sendMessage).not.toHaveBeenCalled();
        expect(startBet).not.toHaveBeenCalled();
    })

    test('startbet from broadcaster with finished round', async () => {
        (requireUser as jest.Mock).mockReturnValueOnce({useBets: true});
        (requireBettingRound as jest.Mock).mockReturnValueOnce({status: 'finished'});
        await processCommands(fakeChannel, {...fakeTags, badges: {broadcaster: 'true'}}, '!e');
        expect(requireBettingRound).toHaveBeenCalledTimes(1);
        expect(sendMessage).toHaveBeenCalledTimes(1);
        expect(startBet).toHaveBeenCalledTimes(1);
    })

    test('startbet from broadcaster with running status', async () => {
        (requireUser as jest.Mock).mockReturnValueOnce({useBets: true});
        (requireBettingRound as jest.Mock).mockReturnValueOnce({status: 'running'});
        await processCommands(fakeChannel, {...fakeTags, badges: {broadcaster: 'true'}}, '!e');
        expect(requireBettingRound).toHaveBeenCalledTimes(1);
        expect(sendMessage).not.toHaveBeenCalled();
        expect(startBet).not.toHaveBeenCalled();
    })

    test('winner with invalid payload', async () => {
        (requireUser as jest.Mock).mockReturnValueOnce({useBets: true});
        (requireBettingRound as jest.Mock).mockReturnValueOnce({status: 'running'});
        await processCommands(fakeChannel, fakeTags, '!g asdf');
        expect(requireBettingRound).toHaveBeenCalledTimes(1);
        expect(sendMessage).not.toHaveBeenCalled();
        expect(startBet).not.toHaveBeenCalled();
    })

    test('winner from non broadcaster', async () => {
        (requireUser as jest.Mock).mockReturnValueOnce({useBets: true});
        (requireBettingRound as jest.Mock).mockReturnValueOnce({status: 'running'});
        await processCommands(fakeChannel, fakeTags, '!g a');
        expect(requireBettingRound).toHaveBeenCalledTimes(1);
        expect(sendMessage).not.toHaveBeenCalled();
        expect(startBet).not.toHaveBeenCalled();
    })

    test('winner with invalid status', async () => {
        (requireUser as jest.Mock).mockReturnValueOnce({useBets: true});
        (requireBettingRound as jest.Mock).mockReturnValueOnce({status: 'betting'});
        await processCommands(fakeChannel, {...fakeTags, badges: {broadcaster: 'true'}}, '!g a');
        expect(requireBettingRound).toHaveBeenCalledTimes(1);
        expect(sendMessage).not.toHaveBeenCalled();
        expect(startBet).not.toHaveBeenCalled();
    })

    test('winner a', async () => {
        (requireUser as jest.Mock).mockReturnValueOnce({useBets: true});
        (requireBettingRound as jest.Mock).mockReturnValueOnce({status: 'running'});
        await processCommands(fakeChannel, {...fakeTags, badges: {broadcaster: 'true'}}, '!g a');
        expect(requireBettingRound).toHaveBeenCalledTimes(1);
        expect(patchBetRound).toHaveBeenCalledTimes(1);
        expect(patchBetRound).toHaveBeenLastCalledWith(1, {result: 'a', status: 'finished'}, true, undefined);
    })

    test('winner b', async () => {
        (requireUser as jest.Mock).mockReturnValueOnce({useBets: true});
        (requireBettingRound as jest.Mock).mockReturnValueOnce({status: 'running'});
        await processCommands(fakeChannel, {...fakeTags, badges: {broadcaster: 'true'}}, '!g b');
        expect(requireBettingRound).toHaveBeenCalledTimes(1);
        expect(patchBetRound).toHaveBeenCalledTimes(1);
        expect(patchBetRound).toHaveBeenLastCalledWith(1, {result: 'b', status: 'finished'}, true, undefined);
    })

    test('betting with invalid payload', async () => {
        (requireUser as jest.Mock).mockReturnValueOnce({useBets: true});
        (requireBettingRound as jest.Mock).mockReturnValueOnce({status: 'betting'});
        await processCommands(fakeChannel, {...fakeTags, badges: {broadcaster: 'true'}}, '!f asdf');
        expect(createBet).not.toHaveBeenCalled();
    })

    test('betting with wrong status', async () => {
        (requireUser as jest.Mock).mockReturnValueOnce({useBets: true});
        (requireBettingRound as jest.Mock).mockReturnValueOnce({status: 'running'});
        await processCommands(fakeChannel, {...fakeTags, badges: {broadcaster: 'true'}}, '!f a');
        expect(createBet).not.toHaveBeenCalled();
    })

    test('betting on a', async () => {
        (requireUser as jest.Mock).mockReturnValueOnce({useBets: true});
        (requireBettingRound as jest.Mock).mockReturnValueOnce({status: 'betting', betters: [], bets: 0, aBets: 0, bBets: 0});
        await processCommands(fakeChannel, {...fakeTags, badges: {broadcaster: 'true'}}, '!f a');
        expect(createBet).toHaveBeenCalledTimes(1);
    })
    
    test('betting on b', async () => {
        (requireUser as jest.Mock).mockReturnValueOnce({useBets: true});
        (requireBettingRound as jest.Mock).mockReturnValueOnce({status: 'betting', betters: [], bets: 0, aBets: 0, bBets: 0});
        await processCommands(fakeChannel, {...fakeTags, badges: {broadcaster: 'true'}}, '!f a');
        expect(createBet).toHaveBeenCalledTimes(1);
    })
    
    test('betting twice creates only one call', async () => {
        (requireUser as jest.Mock).mockReturnValueOnce({useBets: true});
        (requireBettingRound as jest.Mock).mockReturnValueOnce({status: 'betting', betters: [], bets: 0, aBets: 0, bBets: 0});
        await processCommands(fakeChannel, {...fakeTags, badges: {broadcaster: 'true'}}, '!f a');
        expect(createBet).toHaveBeenCalledTimes(1);
        await processCommands(fakeChannel, {...fakeTags, badges: {broadcaster: 'true'}}, '!f a');
        expect(createBet).toHaveBeenCalledTimes(1);
    })
    
    test('betting on x', async () => {
        (requireUser as jest.Mock).mockReturnValueOnce({useBets: true});
        (requireBettingRound as jest.Mock).mockReturnValueOnce({status: 'betting', betters: [], bets: 0, aBets: 0, bBets: 0});
        await processCommands(fakeChannel, {...fakeTags, badges: {broadcaster: 'true'}}, '!f x');
        expect(createBet).not.toHaveBeenCalled();
    })
})