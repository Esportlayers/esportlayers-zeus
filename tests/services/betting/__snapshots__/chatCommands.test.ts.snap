// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getBettingCommands recv 1`] = `
Object {
  "bet": Object {
    "active": true,
    "command": "!f",
    "identifier": "bet",
    "message": "f response",
    "type": "betting_user",
  },
  "startBet": Object {
    "active": true,
    "command": "!e",
    "identifier": "startbet",
    "message": "e response",
    "type": "betting_streamer",
  },
  "winner": Object {
    "active": true,
    "command": "!g",
    "identifier": "betwinner",
    "message": "g response",
    "type": "betting_streamer",
  },
}
`;

exports[`requireUserCommands clear of store 1`] = `
Array [
  Object {
    "active": true,
    "command": "!a",
    "message": "a response",
    "type": "betting_user",
  },
  Object {
    "active": false,
    "command": "!b",
    "message": "b response",
    "type": "betting_user",
  },
  Object {
    "active": true,
    "command": "!c",
    "message": "c response",
    "type": "betting_streamer",
  },
  Object {
    "active": true,
    "command": "!e",
    "identifier": "startbet",
    "message": "e response",
    "type": "betting_streamer",
  },
  Object {
    "active": true,
    "command": "!f",
    "identifier": "bet",
    "message": "f response",
    "type": "betting_user",
  },
  Object {
    "active": true,
    "command": "!g",
    "identifier": "betwinner",
    "message": "g response",
    "type": "betting_streamer",
  },
]
`;

exports[`requireUserCommands test first call without store 1`] = `
Array [
  Object {
    "active": true,
    "command": "!a",
    "message": "a response",
    "type": "betting_user",
  },
  Object {
    "active": false,
    "command": "!b",
    "message": "b response",
    "type": "betting_user",
  },
  Object {
    "active": true,
    "command": "!c",
    "message": "c response",
    "type": "betting_streamer",
  },
  Object {
    "active": true,
    "command": "!e",
    "identifier": "startbet",
    "message": "e response",
    "type": "betting_streamer",
  },
  Object {
    "active": true,
    "command": "!f",
    "identifier": "bet",
    "message": "f response",
    "type": "betting_user",
  },
  Object {
    "active": true,
    "command": "!g",
    "identifier": "betwinner",
    "message": "g response",
    "type": "betting_streamer",
  },
]
`;

exports[`requireUserCommands test second call with store 1`] = `
Array [
  Object {
    "active": true,
    "command": "!a",
    "message": "a response",
    "type": "betting_user",
  },
  Object {
    "active": false,
    "command": "!b",
    "message": "b response",
    "type": "betting_user",
  },
  Object {
    "active": true,
    "command": "!c",
    "message": "c response",
    "type": "betting_streamer",
  },
  Object {
    "active": true,
    "command": "!e",
    "identifier": "startbet",
    "message": "e response",
    "type": "betting_streamer",
  },
  Object {
    "active": true,
    "command": "!f",
    "identifier": "bet",
    "message": "f response",
    "type": "betting_user",
  },
  Object {
    "active": true,
    "command": "!g",
    "identifier": "betwinner",
    "message": "g response",
    "type": "betting_streamer",
  },
]
`;
